### ISSUE TITLE
PR: new website

### ISSUE THREAD

**Issue by showerst** [openstates/openstates#3075]
State: PR

PR launched a new legislative website. I'm going to do some work on it today on the branch tim/2019/pr-2020 but doubt I'll finish it before I take the rest of the week off for the holidays, so if anyone wants to pick it up feel free.

**Comment by froi** [comment]
Hello 👋 

I just saw this repo and issue. I'm from [Code 4 Puerto Rico](https://github.com/code4puertorico) and was introduced to the maintainers of https://sutra.oslpr.org/osl/esutra/.

I'm trying to see if Code 4 Puerto Rico can collaborate with them to get this data in an easy to consume fashion. If you'd like I can report back anything that comes out of our effort.

I'm also willing to help out with this issue if you still want help 😄

**Comment by showerst** [comment]
:wave: hi froi, any help would be great! 

I had been working on a scraper for http://www.tucamarapr.org/dnncamara/web/ActividadLegislativa/TramiteLegislativo.aspx

-- Is https://sutra.oslpr.org/osl/esutra/ an official source? It looks like it will be much easier to scrape.

I'll write up a separate response with what they could provide to make this easier to consume.

**Comment by froi** [comment]
Hey @showerst. 

I've been told that https://sutra.oslpr.org/osl/esutra/ is the official source for the entire legislature. So I would focus on that. I'm trying to see how I can get this data in some machine consumable way. I haven't gotten an answer yet but will keep trying.

http://www.tucamarapr.org/dnncamara/web/ActividadLegislativa/TramiteLegislativo.aspx is just for the state house of representatives.

The state senate has https://www.senado.pr.gov/Pages/Welcome.aspx.

**Comment by showerst** [comment]
Awesome.

RE:Changes they could make -- I'd say there's a few issues with scraping the existing site. Let me know if i've just missed a setting, my spanish is pretty poor  --

1. The pagination is really slow if you up it to 100 items per page, but if you leave it at 10 or 25 we have to make a LOT of requests to get all the bills in a session.
2. Having to make a follow-on request to get all the data (actions, etc) really multiplies the number of request and so our load on the server.

The date range filter is a big help though. If they were going to do something machine readable, having some kind of dump where they put everything available (each bill with sponsors, versions, actions, etc) into one big JSON file would be ideal. (Or a few files, whatever). 

Even if it were only updated nightly or weekly, that would be fine. Then we could just grab that big file, process it, and do a scrape for the few bills with changes since the last time it was published.

**Comment by froi** [comment]
If I get anything from them I'll report back. 

I haven't really taken a look at site but it wouldn't surprise me that it's a pain to scrape. I'll also take a look at it.

**Comment by showerst** [comment]
FYI have a new branch going for this, started from scratch on the new PR site.

https://github.com/openstates/openstates/pull/3181

**Comment by Nosferican** [comment]
Glad to see good progress on that. Just wanted to document that the current scrapper misses the legislator vote level data. The current data only shows the vote tally but not the vote records. For the Senate that is in this awful Flash app http://votaciones.senado.pr.gov that generates PDF with the information. For the House, it is suppose to be at http://www.tucamarapr.org/dnncamara/web/ActividadLegislativa/Votaciones.aspx but I have found a few inconsistencies form time to time.

**Comment by showerst** [comment]
Because of how the new site is structured we already have to make tens of thousands of requests to get all of a session's data. We can't sort by recent so we have to go through all the pages (bills) to look for updated ones. We also have to throttle our scraper speed so we don't get banned, so it takes close to a day to do a full scrape.

I'd like that extra vote data but i'm a little leery of adding another pageview or two to every House bill, that might tip the scraper into taking multi-days to scrape each day's updates, which I'd like to avoid. 

As for the flash into pdf, i have no idea how to even start scraping that without adding in a full embedded chrome instance with flash which we've been avoiding so far since it just balloons the project's requirements. Is there any pattern to urls of the generated PDFs? If they just live somewhere maybe we can get around flash and just scrape them.

**Comment by Nosferican** [comment]
They sure love to make it near impossible to get the data...

So a few ideas to explore... Maybe using the calendar newsfeed (e.g., [example](http://www.tucamarapr.org/dnncamara/web/prensa/calendario.aspx?y=2020&m=6&d=16)) to find the bills that had some update. If that is a good reliable way of identifying bills with activity that should cut down on the having to query every bill to figure out if there was an event or not.

For the Senate, it is my understanding the PDF reports are generated on the fly by the server that listens to the commands issued through the app. The alternative I had opted for was to parse the daily sessions PDF which is a lot easier. I had worked a nice collection of NLP rules for it.

Best outcome would be to keep bugging the legislature to have them machine readable and an API rather than scanned faxed scans like they used to a couple years back and behind a Flash app in 2020.
